2022/src/day16.scala

```diff
@@ -0,0 +1,153 @@
+def parse(xs: String): Input = xs.split("\n").map{ case s"Valve $id has flow rate=$flow; tunnel$_ lead$_ to valve$_ $tunnelsStr" =>
+  val tunnels = tunnelsStr.split(", ").toList
+  Room(id, flow.toInt, tunnels)
+}.toList
+
+case class RoomsInfo(
+  /** map of rooms by id */
+  rooms: Map[Id, Room],
+  /** map from starting room to a map containing the best distance to all other rooms */
+  routes: Map[Id, Map[Id, Int]],
+  /** rooms containing non-zero-flow valves */
+  valves: Set[Id]
+)
+
+// precalculate useful things like pathfinding
+def constructInfo(input: Input): RoomsInfo =
+  val rooms: Map[Id, Room]          = Map.from(for r <- input yield r.id -> r)
+  val valves: Set[Id]               = Set.from(for r <- input if r.flow > 0 yield r.id)
+  val tunnels: Map[Id, List[Id]]    = rooms.mapValues(_.tunnels).toMap
+  val routes: Map[Id, Map[Id, Int]] = (valves + "AA").iterator.map{ id => id -> computeRoutes(id, tunnels) }.toMap
+  RoomsInfo(rooms, routes, valves)
+
+// a modified A-star to calculate the best distance to all rooms rather then the best path to a single room
+def computeRoutes(start: Id, neighbors: Id => List[Id]): Map[Id, Int] =
+
+  case class State(frontier: List[(Id, Int)], scores: Map[Id, Int]):
+
+    private def getScore(id: Id): Int = scores.getOrElse(id, Int.MaxValue) 
+    private def setScore(id: Id, s: Int) = State((id, s + 1) :: frontier, scores + (id -> s))
+
+    def dequeued: (Id, State) =
+      val sorted = frontier.sortBy(_._2)
+      (sorted.head._1, copy(frontier = sorted.tail))
+
+    def considerEdge(from: Id, to: Id): State =
+      val toScore = getScore(from) + 1
+      if toScore >= getScore(to) then this
+      else setScore(to, toScore)
+  end State
+
+  object State:
+    def initial(start: Id) = State(List((start, 0)), Map(start -> 0))
+  
+  def recurse(state: State): State =
+    if state.frontier.isEmpty then
+      state
+    else
+      val (curr, currState) = state.dequeued
+      val newState = neighbors(curr)
+        .foldLeft(currState) { (s, n) =>
+          s.considerEdge(curr, n)
+        }
+      recurse(newState)
+
+  recurse(State.initial(start)).scores
+
+end computeRoutes
+
+
+// find the best path (the order of valves to open) and the total pressure released by taking it
+def bestPath(map: RoomsInfo, start: Id, valves: Set[Id], timeAllowed: Int): Int =
+  // each step involves moving to a room with a useful valve and opening it
+  // we don't need to track each (empty) room in between
+  // we limit our options by only considering the still-closed valves
+  // and `valves` has already culled any room with a flow value of 0 -- no point in considering these rooms!
+
+  def recurse(path: List[Id], valvesLeft: Set[Id], timeLeft: Int, totalValue: Int): Int =
+    // recursively consider all plausible options
+    // we are finished when we no longer have time to reach another valve or all valves are open
+    valvesLeft
+      .flatMap{ id =>
```

Comment:
I was gong to suggest for comprehension here, but that doesnt work for a single flatMap :/

Reply:
unless you extract the result of the Option.when :)

Reply:
```scala
for
  id <- valvesLeft
  t <- locally {
    val current  = path.head
    val distance = map.routes(current)(id)
    timeLeft - distance - 1
  }
  if t > 0
yield
  val value = map.rooms(id).flow * t
  recurse(id :: path, valvesLeft - id, t, totalValue + value)
```
hmm not bad